safedelete/queryset.py

```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.delete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with delete."
-        # TODO: Replace this by bulk update if we can
-        for obj in self.all():
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        with transaction.atomic():
+            all_objs = self.all()
+            all_objs_id_list = list(all_objs.values_list('id', flat=True))
+            current_policy = force_policy or all_objs.model._safedelete_policy
+
+            if current_policy == HARD_DELETE_NOCASCADE:
+                if not all(map(safedelete_utils.can_hard_delete, self.model.objects.filter(id__in=all_objs_id_list))):
+                    current_policy = SOFT_DELETE
+                else:
+                    current_policy = HARD_DELETE
+
+            if current_policy == NO_DELETE:
+                pass
+
+            elif current_policy == SOFT_DELETE and safedelete_utils.is_safedelete_cls(self.model):
+                if is_pre_signalize:
+                    send_pre_delete_signal(all_objs)
+                all_objs.update(**{FIELD_NAME: timezone.now()})
+
+            elif current_policy == HARD_DELETE:
+                if is_pre_signalize:
+                    send_pre_delete_signal(all_objs, soft=False)
+                all_objs.delete()
+
+            elif current_policy == SOFT_DELETE_CASCADE:
+                for obj in all_objs:
+                    related_objs = related_objects(obj, return_as_dict=True)
+
+                    for model, id_list in related_objs.items():
+                        if safedelete_utils.is_safedelete_cls(model) and hasattr(model, FIELD_NAME):
+                            model.objects.filter(id__in=id_list).delete(current_policy)
+
+            self._result_cache = None
+
     delete.alters_data = True
 
-    def undelete(self, force_policy=None):
+    def undelete(self, force_policy=None, is_post_signalize=False):
```

Comment:
rename `is_post_signalize` to something more meaningfull. btw, why did you add this parameter?